src/Microsoft.AspNetCore.Sockets.Common/MessageBatchFormatter.cs

```diff
@@ -0,0 +1,134 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Sockets
+{
+    public static class MessageBatchFormatter
+    {
+        private const byte BatchTerminator = (byte)';';
+        private const byte TextFormatIndicator = (byte)'T';
+        private const byte BinaryFormatIndicator = (byte)'B';
+
+        public static bool TryFormatMessages(IEnumerable<Message> messages, Span<byte> buffer, MessageFormat format, out int bytesWritten)
+        {
+            // Write the format discriminator
+            if (!TryFormatMessageFormat(format, buffer, out var consumedByFormat))
+            {
+                bytesWritten = 0;
+                return false;
+            }
+            buffer = buffer.Slice(consumedByFormat);
+
+            // Write messages
+            var writtenSoFar = 1;
+            foreach (var message in messages)
+            {
+                if (!TryFormatMessage(message, buffer, format, out var writtenForMessage))
+                {
+                    bytesWritten = 0;
+                    return false;
+                }
+                writtenSoFar += writtenForMessage;
+                buffer = buffer.Slice(writtenForMessage);
+            }
+
+            if(buffer.Length < 1)
+            {
+                bytesWritten = 0;
+                return false;
+            }
+            buffer[0] = BatchTerminator;
+            bytesWritten = writtenSoFar + 1;
+            return true;
+        }
+
+        public static bool TryParseMessages(ReadOnlySpan<byte> buffer, out IList<Message> messages, out int bytesConsumed)
+        {
+            if(!TryParseMessageFormat(buffer, out var messageFormat, out var consumedForFormat))
+            {
+                // Batch is missing the prefix
+                bytesConsumed = 0;
+                messages = new List<Message>();
+                return false;
+            }
+            var consumedSoFar = consumedForFormat;
+            buffer = buffer.Slice(consumedForFormat);
+
+            var readMessages = new List<Message>();
+            while (TryParseMessage(buffer, messageFormat, out Message message, out int consumedForMessage))
+            {
+                readMessages.Add(message);
+                consumedSoFar += consumedForMessage;
+                buffer = buffer.Slice(consumedForMessage);
+            }
+
+            if (buffer.Length < 1 || buffer[0] != BatchTerminator)
+            {
+                // Batch is invalid, missing the terminator
+                bytesConsumed = 0;
+                messages = new List<Message>();
+                return false;
+            }
+
+            bytesConsumed = consumedSoFar + 1;
+            messages = readMessages;
+            return true;
+        }
+
+        private static bool TryParseMessageFormat(ReadOnlySpan<byte> buffer, out MessageFormat format, out int bytesConsumed)
+        {
```

Comment:
Yeah, I was trying to figure out the right pattern for this and probably went overkill on this one :)